lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,215 @@
+require 'oci'
+
+module RecordStore
+  class Provider::OracleCloudDNS < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        config = OCI::Config.new
+        config.user = secrets["user"]
+        config.fingerprint = secrets["fingerprint"]
+        config.key_content = secrets["key_content"]
+        config.tenancy = secrets["tenancy"]
+        config.region = secrets["region"]
+        OCI::Dns::DnsClient.new(config: config)
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout) # rubocop:disable Lint/UnusedMethodArgument
+        all_records = []
+        client.get_zone_records(zone, zone_version: zone_version(zone)).each do |response|
+          all_records += response.data.items
+        end
+        all_records.map { |r| build_from_api(r) }.flatten.compact
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets["compartment_id"]).data.map(&:name)
+      end
+
+      private
+
+      # Returns the version of the zone
+      def zone_version(zone)
+        client.get_zone(zone).data.version
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record_rdata,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
```

Comment:
`patch` in APIs is rarely idempotent, good question!

Reply:
If the record exists then it will give me an error from OCI side.
```
OCI::Errors::ServiceError: Found duplicate records in the request: [ test_add_changeset.test.recordstore.io. 600 IN A 10.10.10.42]
```

Reply:
Was just wondering if we're being deliberate about the behaviour here.  Is this what you intend to happen here, @gooallen ?

Reply:
@cttttt Yeah, but I think I better handle if there's an existing record same as what it's added.